File: Medium/C++/Coin_Change(322).cpp
class Solution {
public:
    int coinChange(vector<int> &coins, int amount) {
        
        const int INF = 0x3f3f3f3f;
        int *dp = new int[amount + 1]();
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            dp[i] = INF;
            for (auto &coin : coins) {
                if (i >= coin && dp[i - coin] != INF) {
                    dp[i] = min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount] == INF ? -1 : dp[amount];
    }
};

File: Medium/Python 3/Coin_Change(322).py
class Solution:
    def coinChange(self, coins, amount):
        if not coins or len(coins) == 0 or amount < 0:
            return -1
        
        n = len(coins)
        dp = [sys.maxsize] * (amount + 1)
        dp[0] = 0
        
        for i in range(1, amount + 1):
            for j in range(n):
                last_index = i - coins[j]
                if last_index < 0 or dp[last_index] == sys.maxsize:
                    continue
                dp[i] = min(dp[i], dp[last_index] + 1)
        
        if dp[amount] == sys.maxsize:
            return -1
        
        return dp[amount]
        